collection/collection.go

```diff
@@ -28,22 +28,217 @@ var (
+		// for now we'll hack in an addition to all user's collections, which would
+		// DEFINITELY break cloud
+		if vi, ok := e.Payload.(dsref.VersionInfo); ok {
+			if err := c.set.AddOneAcrossAllCollections(ctx, vi); err != nil {
+				log.Debugw("adding dataset across all collections", "initID", vi.InitID, "err", err)
+			}
+		}
+	case event.ETDatasetPulled:
+		profileID := e.ProfileID
+		if vi, ok := e.Payload.(dsref.VersionInfo); ok && profileID != "" {
+			if err := c.set.AddOneToCollection(ctx, profileID, vi); err != nil {
+				log.Debugw("adding dataset to collection", "profileID", profileID, "initID", vi.InitID, "err", err)
+			}
+		}
+	case event.ETAutomationWorkflowStarted:
+		if evt, ok := e.Payload.(event.WorkflowStartedEvent); ok {
+			err := c.set.UpdateOneAcrossAllCollections(ctx, evt.InitID, func(vi *dsref.VersionInfo) {
+				vi.RunID = evt.RunID
+				vi.RunStatus = "running"
+			})
+			if err != nil {
+				log.Debugw("updating dataset across all collections", "InitID", evt.InitID, "err", err)
+			}
+		}
+	case event.ETAutomationWorkflowStopped:
+		if evt, ok := e.Payload.(event.WorkflowStoppedEvent); ok {
+			err := c.set.UpdateOneAcrossAllCollections(ctx, evt.InitID, func(vi *dsref.VersionInfo) {
+				vi.RunStatus = evt.Status
+			})
+			if err != nil {
+				log.Debugw("updating dataset across all collections", "InitID", evt.InitID, "err", err)
+			}
+		}
+	case event.ETAutomationWorkflowCreated:
+		if wf, ok := e.Payload.(workflow.Workflow); ok {
+			err := c.set.UpdateOneAcrossAllCollections(ctx, wf.InitID, func(vi *dsref.VersionInfo) {
+				vi.WorkflowID = wf.WorkflowID()
+			})
+
+			if err != nil {
+				log.Debugw("updating dataset across all collections", "InitID", wf.InitID, "err", err)
+			}
+		}
+	case event.ETAutomationWorkflowRemoved:
+		if wf, ok := e.Payload.(workflow.Workflow); ok {
+			err := c.set.UpdateOneAcrossAllCollections(ctx, wf.InitID, func(vi *dsref.VersionInfo) {
+				vi.WorkflowID = ""
+			})
+
+			if err != nil {
+				log.Debugw("updating dataset across all collections", "InitID", wf.InitID, "err", err)
+			}
+		}
+	case event.ETFSICreateLink:
+		if link, ok := e.Payload.(event.FSICreateLink); ok {
+			c.set.UpdateOneAcrossAllCollections(ctx, link.InitID, func(vi *dsref.VersionInfo) {
+				vi.FSIPath = link.FSIPath
+			})
+		}
+	case event.ETFSIRemoveLink:
+		if change, ok := e.Payload.(event.FSIRemoveLink); ok {
+			c.set.UpdateOneAcrossAllCollections(ctx, change.InitID, func(vi *dsref.VersionInfo) {
+				vi.FSIPath = ""
+			})
+		}
+	}
+	return nil
+}
+
+// Set maintains lists of datasets, with each list scoped to a user profile.
 type Set interface {
 	List(ctx context.Context, pid profile.ID, lp params.List) ([]dsref.VersionInfo, error)
+	Get(ctx context.Context, pid profile.ID, initID string) (dsref.VersionInfo, error)
 }
 
 // WritableSet is an extension interface on Set that adds methods for
 // adding and removing items
 type WritableSet interface {
```

Comment:
As mentioned when chatting, I'm not sure if we gain anything by having WritableSet be a different interface, rather than just having Set contain methods for adding. If a Set is meant to be read-only, then those methods could just return errors. This suggestion would of course lose compile-time guarantees, causes error handling to be needed at runtime, but that's also true if lots of places in code need to type convert Sets into WritableSets.